From 08bfa159c6e9542b6c7944f71bb396ef44834ac1 Mon Sep 17 00:00:00 2001
From: Derek Weitzel <djw8605@gmail.com>
Date: Tue, 12 Mar 2013 14:41:59 -0500
Subject: [PATCH] Add BATCH_GAHP_CHECK_STATUS_ATTEMPTS parameter for INFN jobs. #3533

BATCH_GAHP_CHECK_STATUS_ATTEMPTS allows the INFN job manager to
be resilliant to failures in job status checks.  The configuration
option is an integer describing the number of failed attempts
allowed before putting the job on hold.
---
 src/condor_gridmanager/infnbatchjob.cpp |   21 +++++++++++++++------
 src/condor_gridmanager/infnbatchjob.h   |    1 +
 src/condor_utils/param_info.in          |   17 +++++++++++++++++
 3 files changed, 33 insertions(+), 6 deletions(-)

diff --git a/src/condor_gridmanager/infnbatchjob.cpp b/src/condor_gridmanager/infnbatchjob.cpp
index 65f7f0e..e3e65b9 100644
--- a/src/condor_gridmanager/infnbatchjob.cpp
+++ b/src/condor_gridmanager/infnbatchjob.cpp
@@ -155,6 +155,7 @@ INFNBatchJob::INFNBatchJob( ClassAd *classad )
 	enteredCurrentRemoteState = time(NULL);
 	lastSubmitAttempt = 0;
 	numSubmitAttempts = 0;
+	numStatusCheckAttempts = 0;
 	remoteSandboxId = NULL;
 	remoteJobId = NULL;
 	lastPollTime = 0;
@@ -665,12 +666,20 @@ void INFNBatchJob::doEvaluateState()
 			}
 			if ( rc != GLOBUS_SUCCESS ) {
 				// unhandled error
-				dprintf( D_ALWAYS,
-						 "(%d.%d) blah_job_status() failed: %s\n",
-						 procID.cluster, procID.proc, gahp->getErrorString() );
-				errorString = gahp->getErrorString();
-				gmState = GM_HOLD;
-				break;
+				numStatusCheckAttempts++;
+				if (numStatusCheckAttempts < param_integer("BATCH_GAHP_CHECK_STATUS_ATTEMPTS", 5)) {
+					// We'll check again soon
+					lastPollTime = time(NULL);
+					gmState = GM_SUBMITTED;
+					break;
+				} else {
+					dprintf( D_ALWAYS,
+							"(%d.%d) blah_job_status() failed: %s\n",
+							procID.cluster, procID.proc, gahp->getErrorString() );
+					errorString = gahp->getErrorString();
+					gmState = GM_HOLD;
+					break;
+				}
 			}
 			ProcessRemoteAd( status_ad );
 			delete status_ad;
diff --git a/src/condor_gridmanager/infnbatchjob.h b/src/condor_gridmanager/infnbatchjob.h
index 6b1fa59..0766452 100644
--- a/src/condor_gridmanager/infnbatchjob.h
+++ b/src/condor_gridmanager/infnbatchjob.h
@@ -70,6 +70,7 @@ class INFNBatchJob : public BaseJob
 	time_t enteredCurrentRemoteState;
 	time_t lastSubmitAttempt;
 	int numSubmitAttempts;
+	int numStatusCheckAttempts;
 	char *batchType;
 	char *remoteSandboxId;
 	char *remoteJobId;
diff --git a/src/condor_utils/param_info.in b/src/condor_utils/param_info.in
index 2e6be92..983ba21 100644
--- a/src/condor_utils/param_info.in
+++ b/src/condor_utils/param_info.in
@@ -10922,3 +10922,20 @@ usage=
 url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
 tags=shadow,baseshadow
 id=612
+
+[BATCH_GAHP_CHECK_STATUS_ATTEMPTS]
+aliases=
+default=5
+version=7.9.5
+range=.*
+state=default
+type=int
+is_macro=false
+reconfig=true
+customization=seldom
+friendly_name=Determines how many failed check status attempts before failing an infn gahp job.
+usage=
+url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
+tags=gridmanager,infnbatchjob
+id=612
+
-- 
1.7.4.1

